src/Microsoft.OData.Core/UriParser/Binders/SelectPathSegmentTokenBinder.cs

```diff
@@ -59,7 +59,18 @@ public static ODataPathSegment ConvertNonTypeTokenToSegment(PathSegmentToken tok
                 return new DynamicPathSegment(tokenIn.Identifier);
             }
 
-            throw new ODataException(ODataErrorStrings.MetadataBinder_PropertyNotDeclared(edmType.FullTypeName(), tokenIn.Identifier));
+
+            var odataError = new ODataError();
+            odataError.ErrorCode = ErrorCodes.PROPERTY_NOT_FOUND_IN_TYPE;
+            odataError.Target = tokenIn.Identifier;
+            ODataErrorDetail errorProperty = new ODataErrorDetail();
+            errorProperty.Target = tokenIn.Identifier;
+            ODataErrorDetail errorType = new ODataErrorDetail();
+            errorType.Target = edmType.FullTypeName();
+            odataError.Details.Add(errorProperty);
+            odataError.Details.Add(errorType);
+
+            throw new ODataErrorException(odataError); 

```

Comment:
ODataErrorException meant to be used for deserialization errors. However, I think it's ok to use in that case too,